bftengine/src/bcstatetransfer/BCStateTran.cpp

```diff
@@ -1932,32 +1924,21 @@ void BCStateTran::processData() {
     // if needed, select a source replica
     //////////////////////////////////////////////////////////////////////////
 
-    bool newSourceReplica = false;
-    const uint64_t diffMilli = ((currentSourceReplica_ == NO_REPLICA) || (currTime < timeMilliCurrentSourceReplica_))
-                               ? 0 : (currTime - timeMilliCurrentSourceReplica_);
-
-    if (ShouldReplaceSourceReplica(badDataFromCurrentSourceReplica, diffMilli)) {
-      if (preferredReplicas_.size() == 0) {
-        if (fs == FetchingState::GettingMissingBlocks) {
-          LOG_DEBUG(STLogger, "Adding all peer replicas to preferredReplicas_ (because preferredReplicas_.size()==0)");
-          // in this case, we will try to use all other replicas
-          SetAllReplicasAsPreferred();
-        } else if (fs == FetchingState::GettingMissingResPages) {
-          LOG_DEBUG(STLogger, "Go to state  GettingCheckpointSummaries (because preferredReplicas_.size()==0)");
-          EnterGettingCheckpointSummariesState();
-          break;  // get out !!
-        }
+    bool newSourceReplica = sourceSelector_.updateSource(badDataFromCurrentSourceReplica, currTime);
+    if (newSourceReplica) {
+      if (fs == FetchingState::GettingMissingResPages) {
+        EnterGettingCheckpointSummariesState();
+        return;
       }
-
-      newSourceReplica = true;
-      currentSourceReplica_ = selectSourceReplica();
-      timeMilliCurrentSourceReplica_ = currTime;
+      LOG_DEBUG(STLogger, "Selected new source replica: " << (sourceSelector_.currentReplica_));
+      metrics_.current_source_replica_.Get().Set(sourceSelector_.currentReplica_);
+      metrics_.preferred_replicas_.Get().Set(sourceSelector_.preferredReplicasToString());
       badDataFromCurrentSourceReplica = false;
       clearAllPendingItemsData();
     }
 
-    Assert(currentSourceReplica_ != NO_REPLICA);
-    Assert(timeMilliCurrentSourceReplica_ != 0);

```

Comment:
Have you intentionally got rid of this assert?

Reply:
> Have you intentionally got rid of this assert?

It's part of the condition in `sourceSelector_.hasSource()`